src/indieweb_utils/webmentions/validate.py

```diff
@@ -0,0 +1,159 @@
+from urllib import parse as url_parse
+from typing import List
+
+from bs4 import BeautifulSoup
+import requests
+
+from ..utils.urls import canonicalize_url
+
+
+class WebmentionValidationError(Exception):
+    def __init__(self, message):
+        self.message = message
+
+
+class WebmentionIsGone(Exception):
+    def __init__(self, message):
+        self.message = message
+
+
+def process_vouch(vouch: str, source: str, vouch_list: List[str]) -> bool:
+    # use vouch to flag webmentions for moderation
+    # see Vouch spec for more: https://indieweb.org/Vouch
+    # by default, webmention should require moderation
+    # if a vouch is valid, webmention does not need moderation
+
+    moderate = True
+
+    if vouch and vouch != "":
+        vouch_domain = url_parse.urlparse(vouch).netloc
+
+        if moderate:
+            # only get domains
+            vouch_list = [v[0] for v in vouch_list]
+
+            if vouch_domain in vouch_list:
+                try:
+                    r = requests.get(vouch)
+                except:
+                    return moderate
+
+                soup = BeautifulSoup(r.text, "lxml")
+
+                # find hyperlink with source
+                # required for a vouch to be valid
+                for anchor in soup.find_all("a"):
+                    if anchor.get("href"):
+                        if anchor["href"] == source:
+                            moderate = False
+
+    return moderate
+
+
+def validate_headers(request_item):
+    if request_item.headers.get("Content-Length"):
+        if int(request_item.headers["Content-Length"]) > 10000000:
+            raise WebmentionValidationError("Source is too large.")
+
+    if "text/html" not in request_item.headers["Content-Type"]:
+        raise WebmentionValidationError("This endpoint only supports HTML webmentions.")
+
+    return True
+
+
+def validate_webmention(source: str, target: str, vouch: str = "", vouch_list: List[str] = []) -> bool:        
+    """
+    Check if a webmention is valid.
+
+    :param source: The source URL of the webmention.
+    :type source: str
+    :param target: The target URL of the webmention.
+    :type target: str
+    :param vouch: The vouch URL of the webmention.
+    :type vouch: str
+    :param vouch_list: A list of vouch domains.
+    :type vouch_list: list
+    :return: Boolean to indicate webmention is valid, boolean
+        stating whether the vouch check has passed.
+    :rtype: bool, bool
+    """
+
+    if source.strip("/") == target.strip("/"):
+        raise WebmentionValidationError("Source and target cannot be the same URL.")
+
+    source_protocol = url_parse.urlparse(source).scheme
+    target_protocol = url_parse.urlparse(target).scheme
+
+    if source_protocol not in ["http", "https"]:
+        raise WebmentionValidationError("Source must contain a URL scheme.")
```

Comment:
It could be a different shceme e.g. ftp or something. Should probably word this "Source must be http or https" scheme.

Reply:
I have revised the error message for that error and the equivalent target error. This change will be in my next commit.